File: src/main/java/my/study/repository/design/pattern/observe/java/SubjectImpl.java
package my.study.repository.design.pattern.observe.java;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Djh on 2018/7/27 14:51
 * E-Mail ：1544579459@qq.com
 */
public class SubjectImpl implements Subject {
    private List<Observe> mObserveList;

    SubjectImpl() {
        mObserveList = new ArrayList<>();
    }


    @Override
    public void register(Observe observe) {
        if (observe == null) {
            throw new RuntimeException("Observe is null!");
        }
        if (!mObserveList.contains(observe)) {
            mObserveList.add(observe);
        }
    }

    @Override
    public void remove(Observe observe) {
        mObserveList.remove(observe);
    }

    @Override
    public void notifyChange() {
        for (Observe observe : mObserveList) {
            observe.update();
        }
    }
}

File: src/main/java/my/study/repository/design/pattern/observe/kotlin/SubjectImpl.kt
package my.study.repository.design.pattern.observe.kotlin

/**
 * @author djh on  2018/10/10 17:01
 * @E-Mail 1544579459@qq.com
 */
class SubjectImpl : Subject {
    
    override val list: MutableList<Observe> = arrayListOf()
    
    override fun registerObserve(observe: Observe) {
        list.add(observe)
    }
    
    override fun unRegisterObserve(observe: Observe) {
        list.remove(observe)
    }
    
    override fun notifyDataChange() {
        list.forEach {
            it.dataUpdate()
        }
    }
}